### DISCUSSION TITLE
About ui.chat_message

### DISCUSSION THREAD

**Discussion by Judy0406-CJ** [zauberzeug/nicegui#2123] [Q&A] [answered]
### Question

The chatapp example is very helpful to understand the UI element and actions, thanks for providing this. Couple of questions,
1. chat_message.avartar doesn't support local file,
2. what's the best way to implement the streaming response
3. it refreshes the whole chat box, is this necessary? Is it possible to only update the last chat_message? Relevant with item2.

Thanks

**Comment by rodja** [comment, marked as answer]
> chat_message.avartar doesn't support local file,

Oh yes. We would be happy to review any pull requests for this.

> what's the best way to implement the streaming response

I'm not sure. "Best way" is quite strong. A simple way would be to replace a ui.markdown or ui.label content with everything from the streaming response you've got so far. Otherwise you would need to build your own ui element which can update wordwise (similar to ui.log_view which updates line-wise).

> it refreshes the whole chat box, is this necessary? Is it possible to only update the last chat_message? Relevant with item2.

Should be possible. But would involve more sophisticated coding.

**Comment by Viacheslav96** [comment]
Has anything changed since? I managed to get a workaround for using local images for avatars using the QChatMessage slot for avatar

**Reply by falkoschindler** [reply]
> Has anything changed since?

I don't think so.

> I managed to get a workaround for using local images for avatars using the QChatMessage slot for avatar.

Would you like to share your solution? 🙂

**Reply by Viacheslav96** [reply]
Sure. I created a custom wrapper for ui.chat_message with the following constructor

```py
class CustomMessage(ui.chat_message):
    def __init__(self, content: st] = None, name: str = None, label: str = None, stamp: str = None, avatar: str = None, sent: bool = False):

        super().__init__(text=content, name=name, label=label, stamp=stamp, sent=sent)

        with self.add_slot("avatar"):
            self.avatar = ui.image(avatar).classes(f"w-12 h-12 {"ml-6" if sent else "mr-6"}")
```

It works nicely, but I need to resize the avatar and add margins manually. I think CSS styles can do it, but I decided to make it simpler

**Reply by Viacheslav96** [reply]
Since last week, I was looking for a way to add local files as backgrounds, and encountered a discussion [[#4322]](https://github.com/zauberzeug/nicegui/discussions/4322#discussioncomment-12191623). By adding a local folder as a static file location, it's possible to have local files as avatars.
